### DISCUSSION TITLE
Execute C# scripts with a button or support #load in Macros

### DISCUSSION THREAD

**Discussion by Nick-C-137** [TabularEditor/TabularEditor3#1330] [Ideas]
### Describe the solution you'd like

Is it possible to create a button in TE3 that executes a C# script on my local machine, stored in a path I specify? It would be awesome to have that ability. This way, I can have my scripts in source control, edit them with an external editor and execute the latest version from within TE3 every time I have to test them. 

Alternatively, you could support the #load operator when saving as a Macro.  

### Goals

1.
2.

**Comment by otykier** [comment]
Hi @Nick-C-137 
This has been discussed before, see https://github.com/TabularEditor/TabularEditor3/discussions/1022 and https://github.com/TabularEditor/TabularEditor3/issues/277.
There are some security concerns associated with loading arbitrary C# code for execution. Today, scripts that are saved as macros are stored in your `%LocalAppData%` folder, from which other users of the machine cannot access them.

We don't have a good solution for this.

In the short term, we could allow specifying another location for saved macros or even using the `#load` directive, as you suggested. However, there would have to be some guarding mechanism, warnings or otherwise, to ensure users do not inadvertently execute code that they didn't author.

**Reply by Nick-C-137** [reply]
Hi again,

Thanks for taking the time to respond! 

Could it be an option to implement the possibility to use the #load directive, but warn the user whenever they try to save a macro with a script that contains this directive? Then you could allow the user to ignore the warning for that specific macro upon save? 

Wouldn't this be ok, as you put the responsibility on the user? A bit similar to when you try to combine queries in Power Query that contain different sources/privacy levels.
